src/main/java/org/elasticsearch/index/search/geo/GeoPolygonFilter.java

```diff
@@ -93,12 +93,24 @@ protected boolean matchDoc(int doc) {
 
         private static boolean pointInPolygon(GeoPoint[] points, double lat, double lon) {
             boolean inPoly = false;
+            double lat0, lon0, lat1, lon1;
+            if (lon < 0) {
+                lon += 360;
+            }
+
+            if (lat <0) {
+                lat += 180;
+            }
 
             for (int i = 1; i < points.length; i++) {
-                if (points[i].lon() < lon && points[i-1].lon() >= lon
-                        || points[i-1].lon() < lon && points[i].lon() >= lon) {
-                    if (points[i].lat() + (lon - points[i].lon()) /
-                            (points[i-1].lon() - points[i].lon()) * (points[i-1].lat() - points[i].lat()) < lat) {
+                lon0 = (points[i-1].lon() < 0) ? points[i-1].lon() + 360.0 : points[i-1].lon();
+                lon1 = (points[i].lon() < 0) ? points[i].lon() + 360.0 : points[i].lon();
+
+                lat0 = (points[i-1].lat() < 0) ? points[i-1].lat() + 180.0 : points[i-1].lat();
+                lat1 = (points[i].lat() < 0) ? points[i].lat() + 180.0 : points[i].lat();
+
+                if (lon1 < lon && lon0 >= lon || lon0 < lon && lon1 >= lon) {
```

Comment:
Actually thinking about this more, I am not confident of my understanding of how this is working (especially given the translation of the non-quadrant 1 points, so that the polygon is no longer enclosed). So a comment here would definitely be appreciated.